counterparty-core/counterpartycore/lib/transaction_helper/transaction_inputs.py

```diff
@@ -0,0 +1,439 @@
+    # manage in an aging cache for any one source address, at any one period
+    # UTXO_P2SH_ENCODING_LOCKS is TTLCache for UTXOs that are used for chaining p2sh encoding
+    #  instead of a simple (txid, vout) key we use [(vin.prevout.hash, vin.prevout.n) for vin tx1.vin]
+    # we cache the make_outkey_vin to avoid having to fetch raw txs too often
+
+    def __init__(self, utxo_locks_max_addresses=None):
+        self.init(utxo_locks_max_addresses)
+
+    def init(self, utxo_locks_max_addresses=None):
+        # config
+        self.utxo_p2sh_encoding_locks = ThreadSafeTTLCache(10000, 180)
+        self.utxo_p2sh_encoding_locks_cache = ThreadSafeTTLCache(1000, 600)
+        self.utxo_locks_max_age = config.UTXO_LOCKS_MAX_AGE
+        self.utxo_locks_max_addresses = utxo_locks_max_addresses or config.UTXO_LOCKS_MAX_ADDRESSES
+        self.utxo_locks_per_address_maxsize = 5000
+
+        self.utxo_locks = None
+        if self.utxo_locks_max_addresses > 0:
+            self.utxo_locks = util.DictCache(self.utxo_locks_max_addresses)
+        print("UTXOLocks initialized")
+        print(self.utxo_locks, self.utxo_locks_max_addresses)
+
+    def make_outkey_vin_txid(self, txid, vout):
+        if (txid, vout) not in self.utxo_p2sh_encoding_locks_cache:
+            txhex = backend.bitcoind.getrawtransaction(txid, verbose=False)
+            self.utxo_p2sh_encoding_locks_cache.set((txid, vout), make_outkey_vin(txhex, vout))
+
+        return self.utxo_p2sh_encoding_locks_cache.get((txid, vout))
+
+    def filter_unspents(self, source, unspent, exclude_utxos):
+        filter_unspents_utxo_locks = []
+        if self.utxo_locks is not None and source in self.utxo_locks:
+            filter_unspents_utxo_locks = self.utxo_locks[source].keys()
+        filter_unspents_p2sh_locks = self.utxo_p2sh_encoding_locks.keys()  # filter out any locked UTXOs to prevent creating transactions that spend the same UTXO when they're created at the same time
+        filtered_unspent = []
+        for output in unspent:
+            if (
+                make_outkey(output) not in filter_unspents_utxo_locks
+                and self.make_outkey_vin_txid(output["txid"], output["vout"])
+                not in filter_unspents_p2sh_locks
+                and (
+                    not exclude_utxos
+                    or not isinstance(exclude_utxos, str)
+                    or f"{output['txid']}:{output['vout']}" not in exclude_utxos.split(",")
+                )
+            ):
+                filtered_unspent.append(output)
+        return filtered_unspent
+
+    def lock_utxos(self, source, inputs):
+        # Lock the source's inputs (UTXOs) chosen for this transaction
+        if self.utxo_locks is not None:
+            if source not in self.utxo_locks:
+                self.utxo_locks[source] = ThreadSafeTTLCache(
+                    self.utxo_locks_per_address_maxsize, self.utxo_locks_max_age
+                )
+            for input in inputs:
```

Comment:
Maybe just `i` is better here so we don't redefine `input`?

Reply:
fixed